Implement backend-agnostic rpc._wait_all_workers() utility

xush6528 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#31888 Implement backend-agnostic rpc._wait_all_workers() utility**
* #31922 Make RRef leak detection always print a warning log

We need a backend-agnostic mechanism to do barrier-like operation before locally destroy RRef context and shutdown RPC Agent.

- Sort worker names.
- Elect the first name as the leader in the ordered worker names.
- Followers reports therir intent to synchronize to the leader.
- Leader also reports to itself, when `_wait_all_workers()` called.
- If all workers report their intent to proceed, leader send the command to every one to proceed.

Differential Revision: [D19290954](https://our.internmc.facebook.com/intern/diff/D19290954/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D19290954/)!

xush6528 committed:
Implement backend-agnostic rpc._wait_all_workers() utility

We need a backend-agnostic mechanism to do barrier-like operation before locally destroy RRef context and shutdown RPC Agent.

- Sort worker names.
- Elect the first name as the leader in the ordered worker names.
- Followers reports therir intent to synchronize to the leader.
- Leader also reports to itself, when `_wait_all_workers()` called.
- If all workers report their intent to proceed, leader send the command to every one to proceed.

Differential Revision: [D19290954](https://our.internmc.facebook.com/intern/diff/D19290954/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D19290954/)!

[ghstack-poisoned]

kostmo commented:
## :pill: CircleCI build failures summary and remediations
As of commit d6ab674c:


* **1/1** failures introduced in this PR


---

<details><summary>This comment was automatically generated by <a href="https://github.com/kostmo/circleci-failure-tracker/tree/master/docs/from-pull-request-comment">Dr. CI</a> (expand for details).</summary>Follow <a href="https://dr.pytorch.org/admin/comments-opt-out.html">this link to opt-out</a> of these comments for your Pull Requests.

Please report bugs/suggestions on the <a href="https://github.com/kostmo/circleci-failure-tracker/issues">GitHub issue tracker</a>.
</details>

jjlilley commented:
What are the main differences vs. the previous reverted D18643137?
(I'm going through the code, they look very similar)

jjlilley commented:
My other aside is that the code is likely going to be more reliable (i.e. less dependent on 5-second timeouts) if we specialize wait_all_workers() on Thrift / process_group backends to use underlying barrier() support.

But this change can happen after wait_all_workers() lands.

xush6528 commented on torch/distributed/rpc/api.py:
```diff
@@ -66,11 +108,55 @@ def _wait_all_workers():
+            except RuntimeError as ex:
+                logger.error(
+                    "{worker_name} failed to respond to 'Shutdown Proceed.' request in {timeout}".format(
+                        worker_name=follower_worker_name,
```
There was a typo on this line. @jjlilley

xush6528 replied:
Also, I need @mrshenli 's help to decide why `test_rref_leak` will be broken by this PR.
This PR adds more work to the `shutdown()` API.
I think the reason is the test does not ensures an RRef leak.

xush6528 commented:
Also, I need @mrshenli 's help to decide why test_rref_leak will be broken by this PR.
This PR adds more work to the shutdown() API.
I think the reason is the test does not ensures an RRef leak.

xush6528 commented:
@jjlilley Sounds good. This PR will be a fall-back implementation.

mrshenli commented:
> I need @mrshenli 's help to decide why `test_rref_leak` will be broken by this PR.

I think `test_rref_leak` is flaky because the test is holding an RRef var alive until after shutdown. More specifically, in the test we have:

```python
    def _test_rref_leak(self):
        ...
        rref = ...
        ...
        rpc.shutdown()
        # exiting scope and de-ref rref
```

Deref of RRef occurs after shutdown, meaning that when shutting down, the RRefContext would still see living forks. This failure is not deterministic because the shutdown message and the delete UserRRef message can be processed in parallel. If the delete UserRRef message was processed first, then there is no failure.

mrshenli commented:
@xush6528 Do we know on which line did `test_rref_leak` crush? Is it indeed because there is no leak sometimes, or is it failed due to other reasons?

mrshenli commented:
Looking at this PR again, I think it should be deterministic as well. All peers will block on `join()` and then hit `_destroy_rref_context`, which throws the error. Or it could be the existing test is also flaky, just didn't show up in the test: when one peer crashes on `_destroy_rref_context`, it deletes its `UserRRef`, which allows it's `rank+1` peer to pass if that peer hasn't run `_destroy_rref_context` yet?

mrshenli commented:
Two thoughts:

1. can we add a print/log to [here](https://github.com/pytorch/pytorch/blob/master/torch/csrc/distributed/rpc/rref_context.cpp#L82) it check if the crash is indeed due to no leak?
2. If 1 is true, I think we can address it by creating the RRef outside of `assertRaisesRegex` and hold it alive after exiting the `with` context, so that the crashing `shutdown` should not trigger UserRRef deletion. Does this make sense?

xush6528 committed:
Update on "Implement backend-agnostic rpc._wait_all_workers() utility"


We need a backend-agnostic mechanism to do barrier-like operation before locally destroy RRef context and shutdown RPC Agent.

- Sort worker names.
- Elect the first name as the leader in the ordered worker names.
- Followers reports therir intent to synchronize to the leader.
- Leader also reports to itself, when `_wait_all_workers()` called.
- If all workers report their intent to proceed, leader send the command to every one to proceed.

Differential Revision: [D19290954](https://our.internmc.facebook.com/intern/diff/D19290954/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D19290954/)!

[ghstack-poisoned]

xush6528 committed:
Update on "Implement backend-agnostic rpc._wait_all_workers() utility"


We need a backend-agnostic mechanism to do barrier-like operation before locally destroy RRef context and shutdown RPC Agent.

- Sort worker names.
- Elect the first name as the leader in the ordered worker names.
- Followers reports therir intent to synchronize to the leader.
- Leader also reports to itself, when `_wait_all_workers()` called.
- If all workers report their intent to proceed, leader send the command to every one to proceed.

Differential Revision: [D19290954](https://our.internmc.facebook.com/intern/diff/D19290954/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D19290954/)!

[ghstack-poisoned]

mrshenli reviewed (approved):
Hopefully #31922 will address the flakiness in `test_rref_leak` by holding the function-local rref var alive till after the exception.

xush6528 commented:
Thanks for all the comments. I will merge this PR and monitor in the following hours.

xush6528 committed:
Update on "Implement backend-agnostic rpc._wait_all_workers() utility"


We need a backend-agnostic mechanism to do barrier-like operation before locally destroy RRef context and shutdown RPC Agent.

- Sort worker names.
- Elect the first name as the leader in the ordered worker names.
- Followers reports therir intent to synchronize to the leader.
- Leader also reports to itself, when `_wait_all_workers()` called.
- If all workers report their intent to proceed, leader send the command to every one to proceed.

Differential Revision: [D19290954](https://our.internmc.facebook.com/intern/diff/D19290954/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D19290954/)!

[ghstack-poisoned]

mrshenli commented:
Hmm, our change didn't fully fix the flakiness. I am reverting this one. Let's debug in that CI docker

https://app.circleci.com/jobs/github/pytorch/pytorch/4167477

mrshenli commented:
So far, it looks like it consistently fails and only fails in py35:

https://app.circleci.com/jobs/github/pytorch/pytorch/4168435
https://app.circleci.com/jobs/github/pytorch/pytorch/4167477
https://app.circleci.com/jobs/github/pytorch/pytorch/4168871
